File: app/app/src/main/java/com/github/wblachowski/camculator/ImageProcessingTask.java
package com.github.wblachowski.camculator;

import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.wblachowski.camculator.processing.result.CompleteResult;
import com.github.wblachowski.camculator.processing.result.ImagePreprocessingResult;

import java.util.List;

public class ImageProcessingTask extends AsyncTask<Object, Void, CompleteResult> {

    private ImageView preview;
    private TextView equationsTextView;

    @Override
    protected CompleteResult doInBackground(Object[] objects) {
        ImageProcessor imageProcessor = (ImageProcessor) objects[0];
        EquationInterpreter equationIntepreter = (EquationInterpreter) objects[1];
        Bitmap bitmap = (Bitmap) objects[2];
        preview = (ImageView) objects[3];
        equationsTextView = (TextView) objects[4];
        ImagePreprocessingResult result = imageProcessor.process(bitmap);
        List<String> equations = equationIntepreter.findEquations(result.getSymbols());
        return new CompleteResult(result, equations);
    }

    protected void onPostExecute(CompleteResult result) {
        preview.setImageBitmap(result.getPreprocessingResult().getBoxesImg());
        equationsTextView.setText(result.getEquations().stream().map(s -> s + '\n').reduce(String::concat).orElse(""));
    }
}

File: app/app/src/main/java/com/github/wblachowski/camculator/ImageProcessingTask.kt
package com.github.wblachowski.camculator

import android.graphics.Bitmap
import android.os.AsyncTask
import android.widget.ImageView
import android.widget.TextView
import com.github.wblachowski.camculator.processing.result.CompleteResult

class ImageProcessingTask : AsyncTask<Any, Void, CompleteResult>() {

    private var preview: ImageView? = null
    private var equationsTextView: TextView? = null

    override fun doInBackground(objects: Array<Any>): CompleteResult {
        val imageProcessor = objects[0] as ImageProcessor
        val equationIntepreter = objects[1] as EquationInterpreter
        val bitmap = objects[2] as Bitmap
        preview = objects[3] as ImageView
        equationsTextView = objects[4] as TextView
        val result = imageProcessor.process(bitmap)
        val equations = equationIntepreter.findEquations(result.symbols)
        return CompleteResult(result, equations)
    }

    override fun onPostExecute(result: CompleteResult) {
        preview!!.setImageBitmap(result.preprocessingResult.boxesImg)
        equationsTextView!!.text = result.equations.stream().map { s -> s + '\n' }.reduce { obj, str -> obj + str }.orElse("")
    }
}